fbw-common/src/wasm/systems/systems/src/communications/receivers.rs

```diff
@@ -0,0 +1,260 @@
+        Self {
+            receive_nav_id: context.get_identifier(format!("{}_IDENT", name)),
+            beep_id: context.get_identifier(format!("ACP_BEEP_IDENT_{}", name)),
+            is_power_supply_powered: false,
+            powered_by,
+            morse: Morse::new(context, name),
+            ok_to_beep: false,
+        }
+    }
+
+    pub fn update(&mut self, context: &UpdateContext, ok_to_beep: bool) {
+        self.ok_to_beep = self.is_power_supply_powered && ok_to_beep;
+        // We keep updating the morse even though the receiver is not powered
+        // because in real life, the signal is external (obviously)
+        self.morse.update(context);
+    }
+}
+
+impl SimulationElement for NavReceiver {
+    fn write(&self, writer: &mut SimulatorWriter) {
+        writer.write(
+            &self.beep_id,
+            if self.ok_to_beep {
+                self.morse.get_state()
+            } else {
+                false
+            },
+        );
+
+        writer.write(&self.receive_nav_id, self.ok_to_beep);
+    }
+
+    fn receive_power(&mut self, buses: &impl ElectricalBuses) {
+        self.is_power_supply_powered = buses.is_powered(self.powered_by);
+    }
+
+    fn accept<T: SimulationElementVisitor>(&mut self, visitor: &mut T) {
+        self.morse.accept(visitor);
+        visitor.visit(self);
+    }
+}
+
+struct Morse {
+    ident_id: VariableIdentifier,
+    ident_new: usize,
+    ident_current: usize,
+    morse: String,
+    beep: bool,
+    duration_short_beep: usize,
+    duration_long_beep: usize,
+    duration_end_of_ident: usize,
+    duration_current: Duration,
+    duration_to_wait: usize,
+}
+
+impl Morse {
+    pub fn new(context: &mut InitContext, name: &str) -> Self {
+        // In milliseconds. For 7 words a minute.
+        // Use the formula here: https://k7mem.com/Keyer_Speed.html
+        let time_base = 171;
+
+        Self {
+            ident_id: context.get_identifier(format!("{}_IDENT_PACKED", name)),
+            ident_new: 0,
+            ident_current: 0,
+            morse: "".to_owned(),
+            beep: false,
+            duration_short_beep: time_base,
+            duration_long_beep: time_base * 3,
+            duration_end_of_ident: time_base * 7,
+            duration_current: Duration::from_millis(0),
+            duration_to_wait: 0,
+        }
+    }
+
+    // From unpack function in file simvar.ts
+    fn unpack(&self, value: usize) -> String {
+        let mut unpacked: String = "".to_owned();
+
+        let mut i: usize = 0;
+        while i < 8 {
+            // pow to returns 0 in the game if big number
+            let power = pow(2, (i % 8) * 6);
+            if power > 0 {
+                let code: usize = (value / power) & 0x3f;
+                if code > 0 {
+                    unpacked.push(char::from_u32((code + 31) as u32).unwrap());
+                }
+            }
+
+            i += 1;
+        }
+
+        unpacked
+    }
```

Comment:
This function can be improved a lot in this way:
```suggestion
    // From unpack function in file simvar.ts
    fn unpack(&self, value: usize) -> impl DoubleEndedIterator<Item = char> {
        (0..8).filter_map(move |i| {
            let code = (value >> i * 6) & 0x3f;
            (code > 0).then(|| char::from_u32((code + 31) as u32).unwrap_or(' '))
        })
    }
```
This has the following advantages:
1. We get rid of `unwrap`
2. We use the power of binary operations
3. We reduce the number of allocations by returning an iterator
4. It's simpler to read and much more "Rustic"

Reply:
Amazing!!!

Coming from C/C++, it's an absolute new way of programming for me haha!